sled-agent/config-reconciler/src/reconciler_task/zones.rs

```diff
@@ -0,0 +1,1119 @@
+//! Module for Omicron zones.
+//!
+//! There is no separate tokio task here; our parent reconciler task owns this
+//! set of zones and is able to mutate it in place during reconciliation.
+
+use crate::SledAgentFacilities;
+use crate::TimeSyncConfig;
+use futures::FutureExt as _;
+use futures::future;
+use id_map::IdMap;
+use id_map::IdMappable;
+use illumos_utils::addrobj::AddrObject;
+use illumos_utils::running_zone::RunCommandError;
+use illumos_utils::running_zone::RunningZone;
+use illumos_utils::zone::AdmError;
+use illumos_utils::zone::Api as _;
+use illumos_utils::zone::DeleteAddressError;
+use illumos_utils::zone::Zones;
+use nexus_sled_agent_shared::inventory::OmicronZoneConfig;
+use nexus_sled_agent_shared::inventory::OmicronZoneType;
+use omicron_common::address::Ipv6Subnet;
+use omicron_uuid_kinds::OmicronZoneUuid;
+use sled_agent_types::time_sync::TimeSync;
+use sled_agent_types::zone_bundle::ZoneBundleCause;
+use sled_storage::config::MountConfig;
+use slog::Logger;
+use slog::info;
+use slog::warn;
+use slog_error_chain::InlineErrorChain;
+use std::net::IpAddr;
+use std::net::Ipv6Addr;
+use std::num::NonZeroUsize;
+use std::str::FromStr as _;
+use std::sync::Arc;
+
+use super::CurrentlyManagedZpools;
+
+#[derive(Debug, Clone)]
+pub enum TimeSyncStatus {
+    NotYetChecked,
+    ConfiguredToSkip,
+    FailedToGetSyncStatus(Arc<TimeSyncError>),
+    TimeSync(TimeSync),
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum TimeSyncError {
+    #[error("no running NTP zone")]
+    NoRunningNtpZone,
+    #[error("multiple running NTP zones - this should never happen!")]
+    MultipleRunningNtpZones,
+    #[error("failed to execute chronyc within NTP zone")]
+    ExecuteChronyc(#[source] RunCommandError),
+    #[error(
+        "failed to parse chronyc tracking output: {reason} (stdout: {stdout:?})"
+    )]
+    FailedToParse { reason: &'static str, stdout: String },
+}
+
+#[derive(Debug)]
+pub(super) struct OmicronZones {
+    zones: IdMap<OmicronZone>,
+    mount_config: Arc<MountConfig>,
+    timesync_config: TimeSyncConfig,
+}
+
+impl OmicronZones {
+    pub(super) fn new(
+        mount_config: Arc<MountConfig>,
+        timesync_config: TimeSyncConfig,
+    ) -> Self {
+        Self { zones: IdMap::default(), mount_config, timesync_config }
+    }
+
+    /// Attempt to shut down any zones that aren't present in `desired_zones`,
+    /// or that weren't present in some prior call but which didn't succeed in
+    /// shutting down and are in a partially-shut-down state.
+    ///
+    /// On failure, returns the number of zones that failed to shut down.
+    pub(super) async fn shut_down_zones_if_needed<T: SledAgentFacilities>(
+        &mut self,
+        desired_zones: &IdMap<OmicronZoneConfig>,
+        sled_agent_facilities: &T,
+        log: &Logger,
+    ) -> Result<(), NonZeroUsize> {
+        self.shut_down_zones_if_needed_impl(
+            desired_zones,
+            sled_agent_facilities,
+            &RealZoneFacilities,
```

Comment:
I'm a big fan of this pattern, and use it in the new trust quorum code. The underlying impl can take any parameter of the correct trait, but we always force the production code to use the real version at compile time without passing it in, which could lead to mistakes.